native/dtrevc3.go

```diff
@@ -0,0 +1,862 @@
+//                                      the columns of VR, in the same order
+//                                      as their eigenvalues.
+// VR is not referenced if side == lapack.LeftEigVec.
+//
+// Complex eigenvectors corresponding to a complex eigenvalue are stored in VL
+// and VR in two consecutive columns, the first holding the real part, and the
+// second the imaginary part.
+//
+// Each eigenvector will be normalized so that the element of largest magnitude
+// has magnitude 1. Here the magnitude of a complex number (x,y) is taken to be
+// |x| + |y|.
+//
+// work must have length at least lwork and lwork must be at least max(1,3*n),
+// otherwise Dtrevc3 will panic. For optimum performance, lwork should be at
+// least n+2*n*nb, where nb is the optimal blocksize.
+//
+// If lwork == -1, instead of performing Dtrevc3, the function only estimates
+// the optimal workspace size based on n and stores it into work[0].
+//
+// Dtrevc3 returns the number of columns in VL and/or VR actually used to store
+// the eigenvectors.
+//
+// Dtrevc3 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dtrevc3(side lapack.EigVecSide, howmny lapack.HowMany, selected []bool, n int, t []float64, ldt int, vl []float64, ldvl int, vr []float64, ldvr int, mm int, work []float64, lwork int) (m int) {
+	switch side {
+	default:
+		panic(badSide)
+	case lapack.RightEigVec, lapack.LeftEigVec, lapack.RightLeftEigVec:
+	}
+	switch howmny {
+	default:
+		panic(badHowMany)
+	case lapack.AllEigVec, lapack.AllEigVecMulQ, lapack.SelectedEigVec:
+	}
+	switch {
+	case n < 0:
+		panic(nLT0)
+	case len(work) < lwork:
+		panic(shortWork)
+	case lwork < max(1, 3*n) && lwork != -1:
+		panic(badWork)
+	}
+	if lwork != -1 {
+		if howmny == lapack.SelectedEigVec {
+			if len(selected) != n {
+				panic("lapack: bad selected length")
+			}
+			// Set m to the number of columns required to store the
+			// selected eigenvectors, and standardize the slice
+			// selected.
+			for j := 0; j < n; {
+				if j == n-1 || t[(j+1)*ldt+j] == 0 {
+					if selected[j] {
+						m++
+					}
+					j++
+					continue
+				}
+				if selected[j] || selected[j+1] {
+					selected[j] = true
+					selected[j+1] = false
+					m += 2
+				}
+				j += 2
+			}
+		} else {
+			m = n
+		}
+		if m > mm {
+			panic("lapack: insufficient number of columns")
+		}
+		checkMatrix(n, n, t, ldt)
+		if (side == lapack.RightEigVec || side == lapack.RightLeftEigVec) && m > 0 {
+			checkMatrix(n, m, vr, ldvr)
+		}
+		if (side == lapack.LeftEigVec || side == lapack.RightLeftEigVec) && m > 0 {
+			checkMatrix(n, m, vl, ldvl)
+		}
+	}
+
+	// Quick return if possible.
+	if n == 0 {
+		work[0] = 1
+		return
+	}
+
+	nb := impl.Ilaenv(1, "DTREVC", string(side)+string(howmny), n, -1, -1, -1)
+
+	// Quick return in case of a workspace query.
+	if lwork == -1 {
+		work[0] = float64(n + 2*n*nb)
+		return m
+	}
+
+	const (
```

Comment:
I've been thinking about a nice place for these constants, L172 is a scenic spot.